test/gtest/asym_mem/test_asymmetric_memory.cc

```diff
@@ -0,0 +1,417 @@
+                                                "init buf");
+            EXPECT_NE(ctxs[i]->init_buf, nullptr);
+            uint8_t *sbuf = (uint8_t*)ctxs[i]->init_buf;
+            for (int j = 0; j < msglen * src_modifier; j++) {
+                sbuf[j] = (uint8_t) 1;
+            }
+            UCC_CHECK(ucc_mc_memcpy(coll->src.info.buffer,
+                                    ctxs[i]->init_buf,
+                                    msglen * src_modifier, src_mem_type,
+                                    UCC_MEMORY_TYPE_HOST));
+
+            ctxs[i]->rbuf_size = msglen * dst_modifier;
+            if (i == root) {
+                UCC_CHECK(ucc_mc_alloc(&ctxs[i]->dst_mc_header,
+                                        ctxs[i]->rbuf_size, dst_mem_type));
+                coll->dst.info.buffer   = ctxs[i]->dst_mc_header->addr;
+                coll->dst.info.count    = (ucc_count_t)ctxs[i]->rbuf_size;
+                coll->dst.info.datatype = UCC_DT_INT8;
+                coll->dst.info.mem_type = dst_mem_type;
+            }
+        }
+    }
+
+    void data_fini()
+    {
+        for (int i = 0; i < ctxs.size(); i++) {
+            gtest_ucc_coll_ctx_t *ctx = ctxs[i];
+            if (!ctx) {
+                continue;
+            }
+            ucc_coll_args_t* coll = ctx->args;
+            UCC_CHECK(ucc_mc_free(ctx->src_mc_header));
+            if (i == coll->root) {
+                UCC_CHECK(ucc_mc_free(ctx->dst_mc_header));
+            }
+            ucc_free(ctx->init_buf);
+            free(coll);
+            free(ctx);
+        }
+        ctxs.clear();
+    }
+
+    bool data_validate()
+    {
+        bool              ret  = true;
+        int               root = 0;
+        uint8_t           result = 1;
+        ucc_memory_type_t dst_mem_type;
+        uint8_t          *rst;
+
+        if (ctxs[0]->args->coll_type == UCC_COLL_TYPE_REDUCE) {
+            result = (uint8_t) ctxs.size();
+        }
+
+        for (int i = 0; i < ctxs.size(); i++) {
+            if (!ctxs[i]) {
+                continue;
+            }
+
+            root = ctxs[i]->args->root;
+
+            if (i == root) {
+                dst_mem_type = ctxs[root]->args->dst.info.mem_type;
+
+                rst = (uint8_t*) ucc_malloc(ctxs[root]->rbuf_size, "validation buf");
+                EXPECT_NE(rst, nullptr);
+
+                UCC_CHECK(ucc_mc_memcpy(rst, ctxs[root]->args->dst.info.buffer,
+                                        ctxs[root]->rbuf_size,
+                                        UCC_MEMORY_TYPE_HOST, dst_mem_type));
+
+                for (int j = 0; j < ctxs[root]->rbuf_size; j++) {
+                    if (result != rst[j]) {
+                        ret = false;
+                        break;
+                    }
+                }
+
+                ucc_free(rst);
+            }
+        }
+        
+        return ret;
+    }
+};
+
+
+class test_asymmetric_memory_v : public ucc::test,
```

Comment:
it's more than that--the buffer allocation, data init, data validate, and etc. all are different because of info_v and because it uses counts/displacements

Reply:
i think it can be merged into one, but i think it would be cleaner to have it separate just so that there isn't an extra layer of if/else for allocation/init/validation/free